src/transforms.rs

```diff
@@ -0,0 +1,254 @@
+    ValueNotAnObject { value: Value },
+
+    #[error("JmespathError: {source}")]
+    JmesPath {
+        #[from]
+        source: jmespath::JmespathError,
+    },
+
+    #[error("serde_json::Error: {source}")]
+    Json {
+        #[from]
+        source: serde_json::Error,
+    },
+}
+
+lazy_static! {
+    pub(crate) static ref TRANSFORM_RUNTIME: Runtime = {
+        let mut runtime = Runtime::new();
+        runtime.register_builtin_functions();
+        runtime.register_function("substr", Box::new(custom_substr()));
+        runtime
+    };
+}
+
+pub(crate) fn compile_transforms(
+    definitions: &HashMap<String, String>,
+) -> Result<HashMap<String, MessageTransform>, TransformError> {
+    let mut transforms = HashMap::new();
+
+    for (k, v) in definitions.iter() {
+        let t = match v.as_str() {
+            "kafka.partition" => MessageTransform::KafkaMetaTransform(KafkaMetaProperty::Partition),
+            "kafka.offset" => MessageTransform::KafkaMetaTransform(KafkaMetaProperty::Offset),
+            "kafka.topic" => MessageTransform::KafkaMetaTransform(KafkaMetaProperty::Topic),
+            "kafka.timestamp" => MessageTransform::KafkaMetaTransform(KafkaMetaProperty::Timestamp),
+            _ => {
+                let expression = TRANSFORM_RUNTIME.compile(v.as_str())?;
+                let expression = expression;
+
+                MessageTransform::ExpressionTransform(expression)
+            }
+        };
+
+        transforms.insert(k.to_owned(), t);
+    }
+
+    Ok(transforms)
+}
+
+fn custom_substr() -> CustomFunction {
+    CustomFunction::new(
+        Signature::new(
+            vec![
+                ArgumentType::String,
+                ArgumentType::String,
+                ArgumentType::String,
+            ],
+            None,
+        ),
+        Box::new(|args: &[Rcvar], _context: &mut Context| {
+            let s = args[0].as_string().unwrap();
+
+            let start = args[1].as_string().unwrap().parse::<usize>().unwrap();
+            let end = args[2].as_string().unwrap().parse::<usize>().unwrap();
+
+            let s: String = s.chars().skip(start).take(end).collect();
+
+            let var = jmespath::Variable::from(serde_json::Value::String(s));
+
+            Ok(Arc::new(var))
+        }),
+    )
+}
+
+pub enum KafkaMetaProperty {
+    Partition,
+    Offset,
+    Topic,
+    Timestamp,
+}
+
+pub enum MessageTransform {
+    KafkaMetaTransform(KafkaMetaProperty),
+    ExpressionTransform(Expression<'static>),
+}
+
+pub(crate) struct TransformContext {
+    transforms: HashMap<String, MessageTransform>,
+}
+
+impl TransformContext {
+    pub(crate) fn new(transforms: HashMap<String, MessageTransform>) -> Self {
+        Self { transforms }
+    }
+
+    pub(crate) fn transform<M>(
+        &self,
+        value: &mut Value,
+        kafka_message: &M,
+    ) -> Result<(), TransformError>
+    where
+        M: Message,
+    {
+        let data = jmespath::Variable::from(value.clone());
```

Comment:
Between this clone of `value` and then the clone of `data` on line 123, I'm a little confused as to why the Variable cannot take ownership of the `value`? Is this just a shortcut for now?

Reply:
I need ownership of `value` becomes I'm mutating it in place just below creating data. You're super right about not needing the clone of `data` below though. `expression.search` is fine with a ref so I've changed that line to

```
let variable = expression.search(&data)?;
```